NtOS Emojipedia - PDA Emoji Messaging For All

san7890 opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

Firstly, apologies for any technical/visual grossities. This is the first full interface I've created using TGUI, and although I tried multiple ways to make it look better, I just wasn't ever able to get it to work (lack of knowledge). I'm also probably missing something big, so apologies for that as well.

Anyways, this PR re-adds Emojipedia (first added in #47177), a tablet function that was lost when we swapped from HTML PDAs to TGUI Tablets. It wasn't a spectacular program, best I could tell in my research all it did was spit out all of the potential emotes to the user's chat with the associated word. This is a bit better than that since it's an actual program application.

One thing I discovered during this PR is that we have three "animated" emotes- however they do not work at all in TGChat, and they make the program look ugly. I pruned down these emotes to only contain one frame in their icon_state, because they just do not work anymore in our current framework.

After some discussion in the PR comments below, I also change the functionality in this PR to allow _everyone_ to send PDA messages! Huzzah! As a hold-over, I think I'll let Curator and Mime start with it shift-start as a reminder of where they once came from (and also because the mime _needs_ it).

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game

![image](https://user-images.githubusercontent.com/34697715/198196019-33bf03f9-1001-4227-8f46-1a6f00883c92.png)

![image](https://user-images.githubusercontent.com/34697715/198196033-f854ad16-a4b4-4fda-a7c9-abc9f06ffdcb.png)

![image](https://user-images.githubusercontent.com/34697715/198196051-010f9cdf-140a-4d7e-b13c-3f08001fb84a.png)

Fixes #67623. This was dropped off for effectively no reason, and it's good to keep functionality going.

<!-- Argue for the merits of your changes and how they benefit the game, especially if they are controversial and/or far reaching. If you can't actually explain WHY what you are doing will improve the game, then it probably isn't good for the game in the first place. -->

## Changelog

<!-- If your PR modifies aspects of the game that can be concretely observed by players or admins you should add a changelog. If your change does NOT meet this description, remove this section. Be sure to properly mark your PRs to prevent unnecessary GBP loss. You can read up on GBP and it's effects on PRs in the tgstation guides for contributors. Please note that maintainers freely reserve the right to remove and add tags should they deem it appropriate. You can attempt to finagle the system all you want, but it's best to shoot for clear communication right off the bat. -->

:cl: san7890, ZephyrTFA
fix: Emojipedia has been restored to tablets! Look for it in your local Nanotrasen Software Hub.
balance: Everyone is now able to send Emojis via PDA Messaging! It's no longer restricted to just mime and curators, so enjoy sending the classics such as :ai:, :1997:, :pslime:, and more!
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

san7890 committed:
NTos Emojipedia - Better Than Momma's To_Chat

Firstly, apologies for any technical/visual grossities. This is the first full interface I've created using TGUI, and although I tried multiple ways to make it look better, I just wasn't ever able to get it to work (lack of knowledge). I'm also probably missing something big, so apologies for that as well.

Anyways, this PR re-adds Emojipedia, a tablet function that was lost when we swapped from HTML PDAs to TGUI Tablets. It wasn't a spectacular program, best I could tell in my research all it did was spit out all of the potential emotes to the user's chat with the associated word. This is a bit better than that since it's an actual program application.

One thing I discovered during this PR is that we have three "animated" emotes- however they do not work at all in TGChat, and they make the program look ugly. I pruned down these emotes to only contain one frame in their icon_state, because they just do not work anymore in our current framework.

I added it to the starting_programs for Mime and Curator since they're the only two that can actually send emoji via PDA messenger (by design). I'm leaving the tablet app for the entire crew to use because:

A) It's harmless. It's just a list of potential emojis.
B) You can still use the emojis in this list in something like OOC chat. This is a bit more of a "meta" application, but it beats looking at the wiki or just having some innate knowledge of emojis.

ZephyrTFA commented:
Think you could make it a grid of the icons and you can hover to see what the key is instead?
I feel like it would look a lot nicer

san7890 commented:
> Think you could make it a grid of the icons and you can hover to see what the key is instead? I feel like it would look a lot nicer

I really, really wanted to try making a grid, something like:
![image](https://user-images.githubusercontent.com/34697715/198180340-efd67ba1-597d-45a3-9939-c6306e72af1d.png)

(imagine every cell has text)

But I tried finagling with TGUI and I just could not figure it out.

ZephyrTFA commented:
> 

let me see if I can make a PR on your branch for this

san7890 commented:
> let me see if I can make a PR on your branch for this

I have skill of noob, thank you so much. Feel free to ping me on Discord `@san7890#7890` if you make the PR and I'm not paying attention or something.

ZephyrTFA committed:
it looks nicer :)

san7890 committed:
Merge pull request #7 from ZephyrTFA/sangrid

Credit to ZephyrTFA <matthew@tfaluc.com>

san7890 committed:
changes ? button to use FA-icon and tooltips

also bump the version

Mothblocks commented on code/modules/modular_computers/file_system/programs/emojipedia.dm:
```diff
@@ -0,0 +1,28 @@
+/datum/computer_file/program/emojipedia
+	filename = "emojipedia"
+	filedesc = "EmojiPedia"
+	category = PROGRAM_CATEGORY_CREW // we want everyone to be able to access this application, since they can use it as a reference for posting in OOC (only mimes and curators can send emoji via PDAs right now)
```
i dont like this thought process regardless

Mothblocks replied:
frankly we need to rethink emojis being mime/curator only. i have just been sending kaomojis in the meantime

san7890 replied:
I can expand emoji-sending to everyone on the station. I thought it was silly too when I discovered it, and it's not documented anywhere player-facing.

Mothblocks replied:
That seems ok I suppose

san7890 replied:
PR updated to allow anyone to message emojis. This is from an Assistant's PDA:

![image](https://user-images.githubusercontent.com/34697715/198406411-7de9e695-badc-4784-994b-08b4e5befc0a.png)

JohnFulpWillard replied:
this is fucked up bruh

san7890 committed:
Everyone can use PDA in messages

stylemistake commented on tgui/packages/tgui/interfaces/NtosEmojipedia.js:
```diff
@@ -0,0 +1,76 @@
+import { useBackend, useSharedState } from '../backend';
+import { Box, Button, Input, Section } from '../components';
+import { NtosWindow } from '../layouts';
+import '../styles/interfaces/Emojipedia.scss';
```
Import it via main.scss file for consistency and faster compile times.

stylemistake commented on tgui/packages/tgui/interfaces/NtosEmojipedia.js:
```diff
@@ -0,0 +1,76 @@
+          padding="0"
+          grid-template-columns="repeat(auto-fill, 32px)"
+          grid-template-rows="repeat(auto-fill, 32px)"
+          grid-gap="4em">
```
These properties, starting with display, aren't supposed to be working on Box and should better stay in CSS, and CSS grid support in IE11 is a little bad, keep that in mind, as long as it works as you expect.

Mothblocks commented on code/modules/modular_computers/file_system/programs/emojipedia.dm:
```diff
@@ -0,0 +1,28 @@
+	// Sort the emoji list so it's easier to find things and we don't have to keep sorting on ui_data since the number of emojis can not change in-game.
+	emoji_list = sortTim(emoji_list, /proc/cmp_text_asc)
+
+/datum/computer_file/program/emojipedia/ui_data(mob_user)
```
Static data would be better than ui_data here, but constant data would be even better

At the very least if this is moved to static data, cache the result of this

san7890 replied:
I'm not entirely sure how to cache the result or use constant data here.

Mothblocks commented on code/modules/modular_computers/file_system/programs/emojipedia.dm:
```diff
@@ -0,0 +1,28 @@
+	for(var/emoji in emoji_list)
+		data["emoji_list"] += list(list(
+			"name" = emoji,
+			"icon64" = icon2base64(icon(EMOJI_SET, emoji)),
```
You should make this a spritesheet instead

Mothblocks replied:
`icon2base64` payloads are kinda huge

san7890 committed:
Spritesheetification and Statification

I have no idea how constant data might work but I did ui_static_data implementation instead.

Spritesheet works now! Apparently, spritesheet importation borks whenever you have a raw number string (very odd), so I had to edit the DMI again.

I also touched everything up to match the suggestions given in the PR. The hover feature is broken :(, but I'll see if we can't bring it back somehow.

san7890 committed:
Merge remote-tracking branch 'upstream/master' into primitive-rock-banging

san7890 committed:
AW HELL YEAH WE'RE COOKING WITH ROCKET FUEL

MY BRAIN SHORT CIRCUITED BUT WE'RE GOOD NOW WE'RE GOOD NOW

san7890 committed:
fixes emoji.dmi for real this time

san7890 committed:
writes a fast unit test because of spritesheet fuckery

maybe if i didn't have to cull icon2base64 i wouldn't have to do this huh? well, we're in the fucked zone now

san7890 commented:
I don't know if it's a limitation of spritesheets or a limitation of how we handle spritesheets via CSS, but basically any icon_states that were just raw numbers would break like such:
![image](https://user-images.githubusercontent.com/34697715/199899145-e6ecc893-dd5b-4adb-a452-7ca56de49b35.png)

With `pensive`'s icon being the last icon in the DMI. I decided that number emojis kinda sucked anyways, so I just removed those and added a unit test to lint against it.

![image](https://user-images.githubusercontent.com/34697715/199899469-86bff7d3-3775-46bc-975c-45211798ace7.png)

:1: going to `:exclamation:`
:1997: going to `:malf:`

![image](https://user-images.githubusercontent.com/34697715/199899518-dc686497-cc9d-4860-9bbe-5f2fa70baec7.png)

Mothblocks commented on code/modules/unit_tests/_unit_tests.dm:
```diff
@@ -191,6 +191,7 @@
 #include "timer_sanity.dm"
 #include "traitor.dm"
 #include "unit_test.dm"
+#include "verify_emoji_names.dm"
 #include "verify_config_tags.dm"
```
ABC order

Mothblocks commented on code/modules/unit_tests/verify_emoji_names.dm:
```diff
@@ -0,0 +1,11 @@
+/// Apparently, spritesheets (or maybe how the CSS backend works) do not respond well to icon_state names that are just pure numbers (which was a behavior in emoji.dmi).
+/// In case we add more emoji, let's just make sure that we don't have any pure numbers in the emoji.dmi file if we ever add more.
+/datum/unit_test/verify_emoji_names
+	priority = 0 // san7890 remove this
```
just use TEST_FOCUS bro

san7890 replied:
I SWORE I REMOVED THIS

san7890 committed:
god fucking dammit

i swore i cleaned that up
